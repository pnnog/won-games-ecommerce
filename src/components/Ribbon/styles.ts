import styled, { css, DefaultTheme } from 'styled-components';
import { RibbonProps, RibbonColors } from '.';
import { darken } from 'polished';
import media from 'styled-media-query';

const WrapperModifiers = {
  color: (theme: DefaultTheme, color: RibbonColors) => css`
    background: ${theme.colors[color]};

    &::before {
      border-left-color: ${darken(0.2, theme.colors[color])};
      border-top-color: ${darken(0.2, theme.colors[color])};
    }
  `,

  small: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xsmall};
    padding: 0 ${theme.spacings.small};
    height: 2.4rem;
    right: -2rem;

    &::before {
      top: 2.4rem;
      border-top-width: 0.7rem;
      border-right-width: 2rem;
    }
  `,
  large: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.small};
    padding: 0 ${theme.spacings.large};
    height: 3.3rem;
    right: -2.5rem;

    &::before {
      top: 3.3rem;
      border-top-width: 1rem;
      border-right-width: 2.6rem;
    }
  `,

  extraLarge: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.large};
    padding: 0 ${theme.spacings.medium};
    height: 4rem;
    right: -3rem;

    &::before {
      top: 4rem;
      border-top-width: 1.3rem;
      border-right-width: 3rem;
    }
  `,
};

export const Wrapper = styled.div<RibbonProps>`
  ${({ theme, color, size }) => css`
    display: flex;
    align-items: center;
    position: absolute;
    top: 5%;
    color: ${theme.colors.white};
    font-weight: ${theme.font.bold};

    &::before {
      content: '';
      position: absolute;
      right: 0;

      border-style: solid;
      border-bottom-width: 1rem;
      border-left-width: 0rem;
      border-right-color: transparent;
      border-bottom-color: transparent;
    }
    ${!!color && WrapperModifiers.color(theme, color)}
    ${!!size && WrapperModifiers[size](theme)}

    ${media.lessThan('medium')`
      right:0;
      ::before{
        display:none; 
    
      }
    `}
  `}
`;
