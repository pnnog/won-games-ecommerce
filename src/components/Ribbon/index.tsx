import * as S from './styles';

export type RibbonColors = 'primary' | 'secondary';
export type RibbonSizes = 'small' | 'large' | 'extraLarge';

export type RibbonProps = {
  children: React.ReactNode;
  color?: RibbonColors;
  size?: RibbonSizes;
};

const Ribbon = ({
  children,
  color = 'primary',
  size = 'small',
}: RibbonProps) => (
  <S.Wrapper color={color} size={size}>
    {children}
  </S.Wrapper>
);

export default Ribbon;
