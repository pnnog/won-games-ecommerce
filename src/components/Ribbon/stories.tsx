import { Meta, Story } from '@storybook/react';
import Ribbon from '.';
import { RibbonProps } from '.';

export default {
  title: 'Ribbon',
  component: Ribbon,

  args: {
    children: 'StoryBook children test',
  },

  argTypes: {
    children: {
      type: 'string',
    },
  },
} as Meta;

export const Default: Story<RibbonProps> = (args) => (
  <div
    style={{
      width: '40rem',
      height: '25rem',
      position: 'relative',
      backgroundColor: 'grey',
    }}
  >
    <Ribbon {...args} />
  </div>
);
