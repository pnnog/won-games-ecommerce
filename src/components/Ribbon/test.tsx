import { screen } from '@testing-library/react';
import { renderWithTheme } from '../../utils/tests/helpers';

import Ribbon from '.';

describe('<Ribbon />', () => {
  it('should render the correctly text', () => {
    renderWithTheme(<Ribbon> Ribbon Test </Ribbon>);
    const element = screen.getByText(/Ribbon Test/i);
    expect(element).toBeInTheDocument();
  });

  it('should render Ribbon with primary color when not is passed', () => {
    renderWithTheme(<Ribbon> Ribbon Test </Ribbon>);
    const element = screen.getByText(/Ribbon Test/i);
    expect(element).toHaveStyleRule('background', '#F231A5');
  });

  it('should render Ribbon with primary color when color is primary', () => {
    renderWithTheme(<Ribbon color="primary"> Ribbon Test </Ribbon>);
    const element = screen.getByText(/Ribbon Test/i);
    expect(element).toHaveStyleRule('background', '#F231A5');
  });

  it('should render Ribbon with secondary color when color is secondary', () => {
    renderWithTheme(<Ribbon color="secondary"> Ribbon Test </Ribbon>);
    const element = screen.getByText(/Ribbon Test/i);
    expect(element).toHaveStyleRule('background', '#3CD3C1');
  });

  it('should render with normal size', () => {
    renderWithTheme(<Ribbon>Ribbon Test</Ribbon>);

    expect(screen.getByText(/Ribbon Test/i)).toHaveStyleRule(
      'height',
      '2.4rem'
    );

    expect(screen.getByText(/Ribbon test/i)).toHaveStyleRule(
      'padding',
      '0 2.4rem'
    );

    expect(screen.getByText(/Ribbon test/i)).toHaveStyleRule(
      'font-size',
      '1.2rem'
    );
  });

  it('should render with large size', () => {
    renderWithTheme(<Ribbon size="large"> Ribbon Test</Ribbon>);

    expect(screen.getByText(/Ribbon test/i)).toHaveStyleRule(
      'height',
      '3.3rem'
    );

    expect(screen.getByText(/Ribbon test/i)).toHaveStyleRule(
      'padding',
      '0 4.0rem'
    );

    expect(screen.getByText(/Ribbon test/i)).toHaveStyleRule(
      'font-size',
      '1.4rem'
    );
  });

  it('should render with extra large size', () => {
    const { container } = renderWithTheme(
      <Ribbon size="extraLarge"> Ribbon Test</Ribbon>
    );

    expect(screen.getByText(/Ribbon test/i)).toHaveStyleRule('height', '4rem');

    expect(screen.getByText(/Ribbon test/i)).toHaveStyleRule(
      'padding',
      '0 3.2rem'
    );

    expect(screen.getByText(/Ribbon test/i)).toHaveStyleRule(
      'font-size',
      '1.8rem'
    );

    expect(container).toMatchSnapshot();
  });
});
