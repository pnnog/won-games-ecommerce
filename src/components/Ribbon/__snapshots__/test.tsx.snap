// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ribbon /> should render with extra large size 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: absolute;
  top: 5%;
  color: #FAFAFA;
  font-weight: 600;
  background: #F231A5;
  font-size: 1.8rem;
  padding: 0 3.2rem;
  height: 4rem;
  right: -3rem;
}

.c0::before {
  content: '';
  position: absolute;
  right: 0;
  border-style: solid;
  border-bottom-width: 1rem;
  border-left-width: 0rem;
  border-right-color: transparent;
  border-bottom-color: transparent;
}

.c0::before {
  border-left-color: #b20b6f;
  border-top-color: #b20b6f;
}

.c0::before {
  top: 4rem;
  border-top-width: 1.3rem;
  border-right-width: 3rem;
}

@media (max-width:768px) {
  .c0 {
    right: 0;
  }

  .c0::before {
    display: none;
  }
}

<div>
   
  <div
    class="c0"
    color="primary"
  >
     Ribbon Test
  </div>
   
</div>
`;
